src/window/utils.rs

```diff
@@ -0,0 +1,86 @@
+use glutin::dpi::PhysicalSize;
+use serde::{Deserialize, Serialize};
+use std::ops::{Div, Mul};
+
+// Maybe this should be independent from serialization?
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Serialize, Deserialize)]
+pub struct Dimensions {
+    pub width: u64,
+    pub height: u64,
+}
+
+impl From<(u64, u64)> for Dimensions {
+    fn from((width, height): (u64, u64)) -> Self {
+        Dimensions { width, height }
+    }
+}
+
+impl From<(f32, f32)> for Dimensions {
+    fn from((width, height): (f32, f32)) -> Self {
+        Dimensions {
+            width: width as u64,
+            height: height as u64,
+        }
+    }
+}
+
+impl From<PhysicalSize<u32>> for Dimensions {
+    fn from(PhysicalSize { width, height }: PhysicalSize<u32>) -> Self {
+        Dimensions {
+            width: width as u64,
+            height: height as u64,
+        }
+    }
+}
+
+impl From<Dimensions> for PhysicalSize<u32> {
+    fn from(Dimensions { width, height }: Dimensions) -> Self {
+        PhysicalSize {
+            width: width as u32,
+            height: height as u32,
+        }
+    }
+}
+
+impl From<Dimensions> for (u64, u64) {
+    fn from(dimensions: Dimensions) -> Self {
+        (dimensions.width, dimensions.height)
+    }
+}
+
+impl From<Dimensions> for (u32, u32) {
+    fn from(dimensions: Dimensions) -> Self {
+        (dimensions.width as u32, dimensions.height as u32)
+    }
+}
+
+impl From<Dimensions> for (i32, i32) {
```

Comment:
It is mostly ready, i'll review it carefully one more time later today, and test this build for a day to catch bugs.

But, i worried that there is this duplication, almost the same `impl From` for u64, u32 and i32. @Kethku maybe there is some better way to write this? i googled, and it can be done using macros, but it looks to obscure for me.

Reply:
I could help writing a macro to avoid this impl repetition

Reply:
Let's try it after this is merged.

Or point me into some docs about how to write it. I found some examples on stackoverflow, but they were not clear enough, so i decided not to bother.

Reply:
I've made an [example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=43ef2f0c10f1a2006977562d30bd426a) how it could be achieved to tuples

But there is some good docs to see using [examples](https://doc.rust-lang.org/reference/macros-by-example.html) and quick refs is needed (to see [token types](https://cheats.rs/#tooling-directives) and [repetition](https://cheats.rs/#macros-attributes))
If you want to see more, there is a good concise [guide](https://danielkeep.github.io/tlborm/book/README.html) to see some tips and tricks